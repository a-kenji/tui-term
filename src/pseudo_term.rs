use ratatui::buffer::Buffer;
use ratatui::layout::Rect;
use termwiz::cell::Intensity;
use termwiz::escape::csi::{Cursor, DecPrivateMode, Edit, Mode, Sgr};
use termwiz::escape::{Action, ControlCode, OperatingSystemCommand, CSI};
use termwiz::surface::CursorShape;

pub mod state;
pub mod termwiz_action;

#[derive(Default, Debug)]
pub struct PseudoTermState {
    area: Rect,
    buf_area: Rect,

    fg: Option<termwiz::color::ColorSpec>,
    bg: Option<termwiz::color::ColorSpec>,
    cursor_style: Option<CursorShape>,
    dec_private_mode: Option<DecPrivateMode>,
    title: Option<String>,
    cursor: (u16, u16),

    dec_auto_wrap: bool,
}

impl PseudoTermState {
    // The default entrypoint
    fn handle_actions(&mut self, actions: &Vec<Action>, area: Rect, buf: &mut Buffer) {
        println!("buf area: {:?}, area: {area:?}", buf.area);
        println!(
            "area right: {:?}, area left: {:?}",
            area.right(),
            area.left()
        );

        self.area = area;
        self.buf_area = buf.area;

        for action in actions {
            println!(
                "action: {:?}, row: {}, col: {}",
                action,
                self.row(),
                self.col()
            );
            match action {
                Action::Print(char) => {
                    buf.get_mut(self.col(), self.row()).set_char(*char);
                    self.advance_col();
                }
                Action::PrintString(string) => {
                    for char in string.chars() {
                        buf.get_mut(self.col(), self.row()).set_char(char);
                        self.advance_col();
                    }
                }
                Action::Control(control) => {
                    self.handle_control(control, area, buf);
                }
                Action::DeviceControl(_) => todo!(),
                Action::OperatingSystemCommand(operating_system_command) => {
                    self.handle_operating_system_command(operating_system_command, area, buf)
                }
                Action::CSI(csi) => self.handle_csi(csi, area, buf),
                Action::Esc(_) => todo!(),
                Action::Sixel(_) => todo!(),
                Action::XtGetTcap(_) => todo!(),
                Action::KittyImage(_) => todo!(),
            }
        }
    }
    fn handle_csi(&mut self, csi: &CSI, _area: Rect, _buf: &mut Buffer) {
        match csi {
            CSI::Sgr(sgr) => self.handle_csi_sgr(sgr, _area, _buf),
            CSI::Cursor(cursor) => self.handle_csi_cursor(cursor, _area, _buf),
            CSI::Edit(edit) => self.handle_csi_edit(edit, _area, _buf),
            CSI::Mode(mode) => self.handle_csi_mode(mode, _area, _buf),
            CSI::Device(_) => todo!(),
            CSI::Mouse(_) => todo!(),
            CSI::Window(_) => todo!(),
            CSI::Keyboard(_) => todo!(),
            CSI::SelectCharacterPath(_, _) => todo!(),
            CSI::Unspecified(_) => todo!(),
        }
    }
    fn handle_csi_mode(&mut self, mode: &Mode, _area: Rect, _buf: &mut Buffer) {
        match mode {
            Mode::SetDecPrivateMode(mode) => {
                self.handle_csi_mode_set_deq_private_mode(mode, _area, _buf)
            }
            Mode::ResetDecPrivateMode(_) => {
                // TODO: implement
            }
            Mode::SaveDecPrivateMode(_) => todo!(),
            Mode::RestoreDecPrivateMode(_) => todo!(),
            Mode::QueryDecPrivateMode(_) => todo!(),
            Mode::SetMode(_) => todo!(),
            Mode::ResetMode(_) => todo!(),
            Mode::QueryMode(_) => todo!(),
            Mode::XtermKeyMode { resource, value } => todo!(),
        }
    }
    fn handle_csi_sgr(&mut self, sgr: &Sgr, _area: Rect, _buf: &mut Buffer) {
        match sgr {
            Sgr::Reset => self.sgr_reset(),
            Sgr::Intensity(intensity) => self.handle_csi_sgr_intensity(intensity, _area, _buf),
            Sgr::Underline(_) => todo!(),
            Sgr::UnderlineColor(_) => todo!(),
            Sgr::Blink(_) => todo!(),
            Sgr::Italic(_) => todo!(),
            Sgr::Inverse(_) => todo!(),
            Sgr::Invisible(_) => todo!(),
            Sgr::StrikeThrough(_) => todo!(),
            Sgr::Font(_) => todo!(),
            Sgr::Foreground(palette) => {
                self.fg = Some(*palette);
            }
            Sgr::Background(_) => todo!(),
            Sgr::Overline(_) => todo!(),
            Sgr::VerticalAlign(_) => todo!(),
        }
    }
    fn handle_csi_sgr_intensity(&mut self, intensity: &Intensity, _area: Rect, _buf: &mut Buffer) {
        match intensity {
            Intensity::Normal => {
                // TODO: implement
            }
            Intensity::Bold => {
                // TODO: implement
            }
            Intensity::Half => {
                // TODO: implement
            }
        }
    }
    fn handle_csi_edit(&mut self, edit: &Edit, _area: Rect, _buf: &mut Buffer) {
        match edit {
            Edit::DeleteCharacter(_) => todo!(),
            Edit::DeleteLine(_) => todo!(),
            Edit::EraseCharacter(_) => todo!(),
            Edit::EraseInLine(_) => {
                // TODO: implement
            }
            Edit::InsertCharacter(_) => todo!(),
            Edit::InsertLine(_) => todo!(),
            Edit::ScrollDown(_) => todo!(),
            Edit::ScrollUp(_) => todo!(),
            Edit::EraseInDisplay(_) => {
                // TODO: implement
            }
            Edit::Repeat(_) => todo!(),
        }
    }
    fn handle_csi_cursor(&mut self, cursor: &Cursor, _area: Rect, _buf: &mut Buffer) {
        match cursor {
            Cursor::BackwardTabulation(_) => todo!(),
            Cursor::TabulationClear(_) => todo!(),
            Cursor::CharacterAbsolute(_) => todo!(),
            Cursor::CharacterPositionAbsolute(_) => todo!(),
            Cursor::CharacterPositionBackward(_) => todo!(),
            Cursor::CharacterPositionForward(_) => todo!(),
            Cursor::CharacterAndLinePosition { line, col } => todo!(),
            Cursor::LinePositionAbsolute(_) => todo!(),
            Cursor::LinePositionBackward(_) => todo!(),
            Cursor::LinePositionForward(_) => todo!(),
            Cursor::ForwardTabulation(_) => todo!(),
            Cursor::NextLine(_) => todo!(),
            Cursor::PrecedingLine(_) => todo!(),
            Cursor::ActivePositionReport { line, col } => todo!(),
            Cursor::RequestActivePositionReport => todo!(),
            Cursor::SaveCursor => todo!(),
            Cursor::RestoreCursor => todo!(),
            Cursor::TabulationControl(_) => todo!(),
            Cursor::Left(amount) => {
                let cols = self.col();
                self.set_col(cols - *amount as u16);
            }
            Cursor::Down(amount) => {
                let rows = self.row();
                self.set_row(rows + *amount as u16);
            }
            Cursor::Right(amount) => {
                let cols = self.col();
                self.set_col(cols + *amount as u16);
            }
            Cursor::Position { line, col } => {
                self.set_col(col.as_zero_based() as u16);
                self.set_row(line.as_zero_based() as u16);
            }
            Cursor::Up(amount) => {
                let rows = self.row();
                self.set_row(rows - *amount as u16);
            }
            Cursor::LineTabulation(_) => todo!(),
            Cursor::SetTopAndBottomMargins { top, bottom } => todo!(),
            Cursor::SetLeftAndRightMargins { left, right } => todo!(),
            Cursor::CursorStyle(cursor_style) => {}
        }
    }
    fn handle_control(&mut self, control_code: &ControlCode, _area: Rect, _buf: &mut Buffer) {
        match control_code {
            ControlCode::Null => todo!(),
            ControlCode::StartOfHeading => todo!(),
            ControlCode::StartOfText => todo!(),
            ControlCode::EndOfText => todo!(),
            ControlCode::EndOfTransmission => todo!(),
            ControlCode::Enquiry => todo!(),
            ControlCode::Acknowledge => todo!(),
            ControlCode::Bell => todo!(),
            ControlCode::Backspace => todo!(),
            ControlCode::HorizontalTab => todo!(),
            ControlCode::LineFeed => {
                let row = self.row();
                self.set_row(row + 1);
            }
            ControlCode::VerticalTab => todo!(),
            ControlCode::FormFeed => todo!(),
            ControlCode::CarriageReturn => {
                self.set_col(0);
            }
            ControlCode::ShiftOut => {
                // ignored
            }
            ControlCode::ShiftIn => {
                // ignored
            }
            ControlCode::DataLinkEscape => todo!(),
            ControlCode::DeviceControlOne => todo!(),
            ControlCode::DeviceControlTwo => todo!(),
            ControlCode::DeviceControlThree => todo!(),
            ControlCode::DeviceControlFour => todo!(),
            ControlCode::NegativeAcknowledge => todo!(),
            ControlCode::SynchronousIdle => todo!(),
            ControlCode::EndOfTransmissionBlock => todo!(),
            ControlCode::Cancel => todo!(),
            ControlCode::EndOfMedium => todo!(),
            ControlCode::Substitute => todo!(),
            ControlCode::Escape => todo!(),
            ControlCode::FileSeparator => todo!(),
            ControlCode::GroupSeparator => todo!(),
            ControlCode::RecordSeparator => todo!(),
            ControlCode::UnitSeparator => todo!(),
            ControlCode::BPH => todo!(),
            ControlCode::NBH => todo!(),
            ControlCode::IND => todo!(),
            ControlCode::NEL => todo!(),
            ControlCode::SSA => todo!(),
            ControlCode::ESA => todo!(),
            ControlCode::HTS => todo!(),
            ControlCode::HTJ => todo!(),
            ControlCode::VTS => todo!(),
            ControlCode::PLD => todo!(),
            ControlCode::PLU => todo!(),
            ControlCode::RI => todo!(),
            ControlCode::SS2 => todo!(),
            ControlCode::SS3 => todo!(),
            ControlCode::DCS => todo!(),
            ControlCode::PU1 => todo!(),
            ControlCode::PU2 => todo!(),
            ControlCode::STS => todo!(),
            ControlCode::CCH => todo!(),
            ControlCode::MW => todo!(),
            ControlCode::SPA => todo!(),
            ControlCode::EPA => todo!(),
            ControlCode::SOS => todo!(),
            ControlCode::SCI => todo!(),
            ControlCode::CSI => todo!(),
            ControlCode::ST => todo!(),
            ControlCode::OSC => todo!(),
            ControlCode::PM => todo!(),
            ControlCode::APC => todo!(),
        }
    }
    fn handle_csi_mode_set_deq_private_mode(
        &mut self,
        mode: &DecPrivateMode,
        _area: Rect,
        _buf: &mut Buffer,
    ) {
        match mode {
            DecPrivateMode::Code(dec_private_mode) => {
                match dec_private_mode {
                    termwiz::escape::csi::DecPrivateModeCode::ApplicationCursorKeys => todo!(),
                    termwiz::escape::csi::DecPrivateModeCode::DecAnsiMode => todo!(),
                    termwiz::escape::csi::DecPrivateModeCode::Select132Columns => todo!(),
                    termwiz::escape::csi::DecPrivateModeCode::SmoothScroll => todo!(),
                    termwiz::escape::csi::DecPrivateModeCode::ReverseVideo => todo!(),
                    termwiz::escape::csi::DecPrivateModeCode::OriginMode => todo!(),
                    termwiz::escape::csi::DecPrivateModeCode::AutoWrap => {
                        self.dec_auto_wrap = true;
                    }
                    termwiz::escape::csi::DecPrivateModeCode::AutoRepeat => todo!(),
                    termwiz::escape::csi::DecPrivateModeCode::StartBlinkingCursor => todo!(),
                    termwiz::escape::csi::DecPrivateModeCode::ShowCursor => todo!(),
                    termwiz::escape::csi::DecPrivateModeCode::ReverseWraparound => todo!(),
                    termwiz::escape::csi::DecPrivateModeCode::LeftRightMarginMode => todo!(),
                    termwiz::escape::csi::DecPrivateModeCode::SixelDisplayMode => todo!(),
                    termwiz::escape::csi::DecPrivateModeCode::MouseTracking => todo!(),
                    termwiz::escape::csi::DecPrivateModeCode::HighlightMouseTracking => todo!(),
                    termwiz::escape::csi::DecPrivateModeCode::ButtonEventMouse => todo!(),
                    termwiz::escape::csi::DecPrivateModeCode::AnyEventMouse => todo!(),
                    termwiz::escape::csi::DecPrivateModeCode::FocusTracking => {
                        // TODO: this should be handled by the terminal
                    }
                    termwiz::escape::csi::DecPrivateModeCode::Utf8Mouse => todo!(),
                    termwiz::escape::csi::DecPrivateModeCode::SGRMouse => todo!(),
                    termwiz::escape::csi::DecPrivateModeCode::SGRPixelsMouse => todo!(),
                    termwiz::escape::csi::DecPrivateModeCode::XTermMetaSendsEscape => todo!(),
                    termwiz::escape::csi::DecPrivateModeCode::XTermAltSendsEscape => todo!(),
                    termwiz::escape::csi::DecPrivateModeCode::SaveCursor => todo!(),
                    termwiz::escape::csi::DecPrivateModeCode::ClearAndEnableAlternateScreen => todo!(),
                    termwiz::escape::csi::DecPrivateModeCode::EnableAlternateScreen => todo!(),
                    termwiz::escape::csi::DecPrivateModeCode::OptEnableAlternateScreen => todo!(),
                    termwiz::escape::csi::DecPrivateModeCode::BracketedPaste => {
                        // TODO: this should be handled by the terminal
                    }
                    termwiz::escape::csi::DecPrivateModeCode::UsePrivateColorRegistersForEachGraphic => todo!(),
                    termwiz::escape::csi::DecPrivateModeCode::SynchronizedOutput => todo!(),
                    termwiz::escape::csi::DecPrivateModeCode::MinTTYApplicationEscapeKeyMode => todo!(),
                    termwiz::escape::csi::DecPrivateModeCode::SixelScrollsRight => todo!(),
                    termwiz::escape::csi::DecPrivateModeCode::Win32InputMode => todo!(),
                }
            }
            DecPrivateMode::Unspecified(_) => todo!(),
        }
    }
    fn handle_operating_system_command(
        &mut self,
        operating_system_command: &OperatingSystemCommand,
        _area: Rect,
        _buf: &mut Buffer,
    ) {
        match operating_system_command {
            OperatingSystemCommand::SetIconNameAndWindowTitle(title) => {
                // TODO: set icon name
                self.title = Some(title.into());
            }
            OperatingSystemCommand::SetWindowTitle(_) => todo!(),
            OperatingSystemCommand::SetWindowTitleSun(_) => todo!(),
            OperatingSystemCommand::SetIconName(_) => todo!(),
            OperatingSystemCommand::SetIconNameSun(_) => todo!(),
            OperatingSystemCommand::SetHyperlink(_) => todo!(),
            OperatingSystemCommand::ClearSelection(_) => todo!(),
            OperatingSystemCommand::QuerySelection(_) => todo!(),
            OperatingSystemCommand::SetSelection(_, _) => todo!(),
            OperatingSystemCommand::SystemNotification(_) => todo!(),
            OperatingSystemCommand::ITermProprietary(_) => todo!(),
            OperatingSystemCommand::FinalTermSemanticPrompt(_) => todo!(),
            OperatingSystemCommand::ChangeColorNumber(_) => todo!(),
            OperatingSystemCommand::ChangeDynamicColors(_, _) => todo!(),
            OperatingSystemCommand::ResetDynamicColor(_) => todo!(),
            OperatingSystemCommand::CurrentWorkingDirectory(_) => todo!(),
            OperatingSystemCommand::ResetColors(_) => todo!(),
            OperatingSystemCommand::RxvtExtension(_) => todo!(),
            OperatingSystemCommand::Unspecified(_) => todo!(),
        }
    }

    /// Sets sgr attributes to their default values.
    fn sgr_reset(&mut self) {
        self.fg = None;
        self.bg = None;
    }
    fn row(&self) -> u16 {
        self.cursor.0
    }
    fn col(&self) -> u16 {
        self.cursor.1
    }
    fn row_mut(&mut self) -> u16 {
        self.cursor.0
    }
    fn col_mut(&mut self) -> u16 {
        self.cursor.1
    }
    fn set_col(&mut self, col: u16) {
        self.cursor.1 = col;
    }
    fn set_row(&mut self, row: u16) {
        self.cursor.0 = row;
    }
    fn advance_col(&mut self) {
        if self.dec_auto_wrap && self.col() >= self.area.right() - 1 {
            self.advance_row();
            self.cursor.1 = 0;
        } else {
            self.cursor.1 += 1;
        }
    }
    fn advance_row(&mut self) {
        if self.row() >= self.area.bottom() - 1 {
            println!("Buttom reached");
        } else {
            self.cursor.0 += 1;
        }
    }
}
