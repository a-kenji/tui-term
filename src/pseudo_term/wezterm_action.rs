use ratatui::buffer::Buffer;
use ratatui::layout::Rect;
use ratatui::widgets::Widget;
use termwiz::escape::Action;

pub struct PseudoTerm<'a> {
    actions: &'a Vec<Action>,
}

impl<'a> PseudoTerm<'a> {
    pub fn new(actions: &'a Vec<Action>) -> Self {
        PseudoTerm { actions }
    }
}

impl Widget for PseudoTerm<'_> {
    fn render(self, area: Rect, buf: &mut Buffer) {
        // let mut row = area.height;
        // let mut col = area.width;
        let mut row = area.height;
        let mut col = area.width;

        for action in self.actions {
            match action {
                Action::Print(char) => {
                    buf.get_mut(col, row).set_char(*char);
                    col += 1;
                }
                Action::PrintString(string) => {
                    for char in string.chars() {
                        buf.get_mut(col, row).set_char(char);
                    }
                }
                Action::Control(control) => match control {
                    termwiz::escape::ControlCode::Null => todo!(),
                    termwiz::escape::ControlCode::StartOfHeading => todo!(),
                    termwiz::escape::ControlCode::StartOfText => todo!(),
                    termwiz::escape::ControlCode::EndOfText => todo!(),
                    termwiz::escape::ControlCode::EndOfTransmission => todo!(),
                    termwiz::escape::ControlCode::Enquiry => todo!(),
                    termwiz::escape::ControlCode::Acknowledge => todo!(),
                    termwiz::escape::ControlCode::Bell => todo!(),
                    termwiz::escape::ControlCode::Backspace => todo!(),
                    termwiz::escape::ControlCode::HorizontalTab => {
                        // Move to next tab character
                    }
                    termwiz::escape::ControlCode::LineFeed => {
                        row += 1;
                    }
                    termwiz::escape::ControlCode::VerticalTab => todo!(),
                    termwiz::escape::ControlCode::FormFeed => todo!(),
                    termwiz::escape::ControlCode::CarriageReturn => {
                        col = 0;
                    }
                    termwiz::escape::ControlCode::ShiftOut => todo!(),
                    termwiz::escape::ControlCode::ShiftIn => todo!(),
                    termwiz::escape::ControlCode::DataLinkEscape => todo!(),
                    termwiz::escape::ControlCode::DeviceControlOne => todo!(),
                    termwiz::escape::ControlCode::DeviceControlTwo => todo!(),
                    termwiz::escape::ControlCode::DeviceControlThree => todo!(),
                    termwiz::escape::ControlCode::DeviceControlFour => todo!(),
                    termwiz::escape::ControlCode::NegativeAcknowledge => todo!(),
                    termwiz::escape::ControlCode::SynchronousIdle => todo!(),
                    termwiz::escape::ControlCode::EndOfTransmissionBlock => todo!(),
                    termwiz::escape::ControlCode::Cancel => todo!(),
                    termwiz::escape::ControlCode::EndOfMedium => todo!(),
                    termwiz::escape::ControlCode::Substitute => todo!(),
                    termwiz::escape::ControlCode::Escape => todo!(),
                    termwiz::escape::ControlCode::FileSeparator => todo!(),
                    termwiz::escape::ControlCode::GroupSeparator => todo!(),
                    termwiz::escape::ControlCode::RecordSeparator => todo!(),
                    termwiz::escape::ControlCode::UnitSeparator => todo!(),
                    termwiz::escape::ControlCode::BPH => todo!(),
                    termwiz::escape::ControlCode::NBH => todo!(),
                    termwiz::escape::ControlCode::IND => todo!(),
                    termwiz::escape::ControlCode::NEL => todo!(),
                    termwiz::escape::ControlCode::SSA => todo!(),
                    termwiz::escape::ControlCode::ESA => todo!(),
                    termwiz::escape::ControlCode::HTS => todo!(),
                    termwiz::escape::ControlCode::HTJ => todo!(),
                    termwiz::escape::ControlCode::VTS => todo!(),
                    termwiz::escape::ControlCode::PLD => todo!(),
                    termwiz::escape::ControlCode::PLU => todo!(),
                    termwiz::escape::ControlCode::RI => todo!(),
                    termwiz::escape::ControlCode::SS2 => todo!(),
                    termwiz::escape::ControlCode::SS3 => todo!(),
                    termwiz::escape::ControlCode::DCS => todo!(),
                    termwiz::escape::ControlCode::PU1 => todo!(),
                    termwiz::escape::ControlCode::PU2 => todo!(),
                    termwiz::escape::ControlCode::STS => todo!(),
                    termwiz::escape::ControlCode::CCH => todo!(),
                    termwiz::escape::ControlCode::MW => todo!(),
                    termwiz::escape::ControlCode::SPA => todo!(),
                    termwiz::escape::ControlCode::EPA => todo!(),
                    termwiz::escape::ControlCode::SOS => todo!(),
                    termwiz::escape::ControlCode::SCI => todo!(),
                    termwiz::escape::ControlCode::CSI => todo!(),
                    termwiz::escape::ControlCode::ST => todo!(),
                    termwiz::escape::ControlCode::OSC => todo!(),
                    termwiz::escape::ControlCode::PM => todo!(),
                    termwiz::escape::ControlCode::APC => todo!(),
                },
                Action::DeviceControl(_) => {}
                Action::OperatingSystemCommand(_) => {}
                Action::CSI(_) => {}
                Action::Esc(_) => {}
                Action::Sixel(_) => {}
                Action::XtGetTcap(_) => {}
                Action::KittyImage(_) => {}
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use ratatui::backend::TestBackend;
    use ratatui::Terminal;

    use super::*;

    #[test]
    fn empty_actions() {
        let backend = TestBackend::new(80, 24);
        let mut terminal = Terminal::new(backend).unwrap();
        let actions = vec![];
        let pseudo_term = PseudoTerm::new(&actions);
        terminal
            .draw(|f| {
                f.render_widget(pseudo_term, f.size());
            })
            .unwrap();
        let view = terminal.backend().to_string();
        insta::assert_snapshot!(view);
    }

    #[test]
    fn simple_ls() {
        let backend = TestBackend::new(80, 24);
        let mut terminal = Terminal::new(backend).unwrap();
        let actions = vec![];
        let pseudo_term = PseudoTerm::new(&actions);
        terminal
            .draw(|f| {
                f.render_widget(pseudo_term, f.size());
            })
            .unwrap();
        let view = terminal.backend().to_string();
        insta::assert_snapshot!(view);
    }
}
