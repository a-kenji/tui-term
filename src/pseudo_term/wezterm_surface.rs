use ratatui::buffer::Buffer;
use ratatui::layout::Rect;
use ratatui::widgets::Widget;
use termwiz::escape::Action;
use termwiz::surface::Surface;

pub struct PseudoTerm<'a> {
    surface: &'a Surface,
}

impl<'a> PseudoTerm<'a> {
    pub fn new(surface: &'a Surface) -> Self {
        PseudoTerm { surface }
    }
}

impl Widget for PseudoTerm<'_> {
    fn render(self, area: Rect, buf: &mut Buffer) {
        // let mut row = area.height;
        // let mut col = area.width;
        let mut row = 0;
        let mut col = 0;

        let surface = self.surface;

        if surface.has_changes(0) {
            panic!("Surface has changes!");
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn simple_ls_output() {
        let actions = vec![];
    }
}
