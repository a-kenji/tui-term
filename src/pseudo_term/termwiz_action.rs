use ratatui::buffer::Buffer;
use ratatui::layout::Rect;
use ratatui::widgets::Widget;
use termwiz::escape::Action;

use crate::pseudo_term::PseudoTermState;

pub struct PseudoTerm<'a> {
    actions: &'a Vec<Action>,
}

impl<'a> PseudoTerm<'a> {
    pub fn new(actions: &'a Vec<Action>) -> Self {
        PseudoTerm { actions }
    }
}

impl Widget for PseudoTerm<'_> {
    fn render(self, area: Rect, buf: &mut Buffer) {
        let mut state = PseudoTermState::default();
        state.handle_actions(self.actions, area, buf);
    }
}

#[cfg(test)]
mod tests {
    use ratatui::backend::TestBackend;
    use ratatui::Terminal;

    use super::*;

    #[test]
    fn empty_actions() {
        let backend = TestBackend::new(80, 24);
        let mut terminal = Terminal::new(backend).unwrap();
        let actions = vec![];
        let pseudo_term = PseudoTerm::new(&actions);
        terminal
            .draw(|f| {
                f.render_widget(pseudo_term, f.size());
            })
            .unwrap();
        let view = terminal.backend().to_string();
        insta::assert_snapshot!(view);
    }

    #[test]
    fn simple_ls() {
        let backend = TestBackend::new(80, 24);
        let simple_ls = include_bytes!("../../test/typescript/simple_ls.typescript");
        let mut terminal = Terminal::new(backend).unwrap();
        let mut parser = termwiz::escape::parser::Parser::new();
        let actions = parser.parse_as_vec(simple_ls);
        let pseudo_term = PseudoTerm::new(&actions);
        terminal
            .draw(|f| {
                f.render_widget(pseudo_term, f.size());
            })
            .unwrap();
        let view = terminal.backend().to_string();
        insta::assert_snapshot!(view);
    }
    #[test]
    fn vttest_02_01() {
        let backend = TestBackend::new(80, 24);
        let simple_ls = include_bytes!("../../test/typescript/vttest_02_01.typescript");
        let mut terminal = Terminal::new(backend).unwrap();
        let mut parser = termwiz::escape::parser::Parser::new();
        let actions = parser.parse_as_vec(simple_ls);
        let pseudo_term = PseudoTerm::new(&actions);
        terminal
            .draw(|f| {
                f.render_widget(pseudo_term, f.size());
            })
            .unwrap();
        let view = terminal.backend().to_string();
        insta::assert_snapshot!(view);
    }
    #[test]
    fn vttest_02_02() {
        let backend = TestBackend::new(80, 24);
        let simple_ls = include_bytes!("../../test/typescript/vttest_02_02.typescript");
        let mut terminal = Terminal::new(backend).unwrap();
        let mut parser = termwiz::escape::parser::Parser::new();
        let actions = parser.parse_as_vec(simple_ls);
        let pseudo_term = PseudoTerm::new(&actions);
        terminal
            .draw(|f| {
                f.render_widget(pseudo_term, f.size());
            })
            .unwrap();
        let view = terminal.backend().to_string();
        insta::assert_snapshot!(view);
    }
}
