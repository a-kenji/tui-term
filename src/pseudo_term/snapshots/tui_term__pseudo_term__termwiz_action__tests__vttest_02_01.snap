---
source: src/pseudo_term/termwiz_action.rs
expression: view
---
"********************************************************************************"
"********************************************************************************"
"********************************************************************************"
"********************************************************************************"
"This should be three identical lines of *'s completely filling                  "
"the top of the screen without any empty lines between.                          "
"(Test of WRAP AROUND mode setting.)                                             "
"Push <RETURN>                                                                   "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "

